---
title: "Summarizing bigWig files"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Summarizing bigWig files}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
  \usepackage[utf8]{inputenc}
---

```{r include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  warning = FALSE,
  message = FALSE,
  fig.width = 7,
  fig.height = 6
)
```

```{r setup, include = FALSE}
library(ggplot2)
library(rtracklayer)
library(GenomicRanges)
library(wigglescout)

h33_chip <- system.file("extdata", "sample_H33_ChIP.bw", package = "wigglescout")
h3k9me3_chip <- system.file("extdata", "sample_H3K9me3_ChIP.bw", package = "wigglescout")
input_chip <- system.file("extdata", "sample_Input.bw", package = "wigglescout")
genes <- system.file("extdata", "sample_genes_mm9.bed", package = "wigglescout")
chromhmm <- system.file("extdata", "sample_chromhmm.bed", package = "wigglescout")

locus <- GRanges(seqnames = "chr15", IRanges(102600000, 103100000))
```

# Calculating functions

This vignette uses the bundled data to showcase these functions. Please refer
to the quick start guide for more information about the bundled data.

bigWig data can be summarized genome-wide using calculation functions that
leverage `rtracklayer` for handling bigWig files (and `furrr` to boost
performance for large use cases): `bw_bins`, `bw_loci`, `bw_heatmap` or
`bw_profile`.

## Genome-wide analysis

`bw_bins` function returns a `GRanges` object where each entry is a locus of 
length `bin_size` and its score is the mean coverage.

**Note:** Small bin sizes (< 5000bp) take longer to run.

```{r}
# Several bwfiles can be provided at once
bw_bins(c(h33_chip, input_chip),
        bin_size = 2000,
        genome = "mm9",
        selection = locus)
```

It is also possible to provide a list of bigWig files to be used as background
to normalize the bin values to. For instance, in the previous case, one could
want to use the input values to normalize the H3.3 ChIP data:

```{r}
bw_bins(h33_chip, bg_bwfiles = input_chip,
        bin_size = 2000,
        genome = "mm9",
        selection = locus)
```
If you want to get the log2 ratio sample / input, select `norm_mode = "log2fc"`:

```{r}
bw_bins(h33_chip, bg_bwfiles = input_chip,
        bin_size = 2000,
        genome = "mm9",
        selection = locus,
        norm_mode = "log2fc")
```

`norm_mode` is set to `"fc"` by default, which means it shows fold change 
sample / input, if input (background) is provided.

### Supported genomes

`bw_bins` function is based on `tileGenome` function from `GenomicRanges`, which
requires information about chromosome length. This is obtained relying on
function `Seqinfo` from `GenomeInfoDb` package. So, in theory, any genome ID
that would return a `Seqinfo` object by doing `Seqinfo(genome="your_genome")`
should work.

## Locus-based analysis

These functions are useful when you have some genomic annotations you want to
look at, rather than general genome-wide trends.

### Individual loci

To calculate values on a *locus* specific manner you can use `bw_loci` function.
For instance, we can look at the genes at this region:

```{r}
bw_loci(c(h33_chip, input_chip), loci = genes)
```

Many of the options available for `bw_bins` are shared with `bw_loci`, so you can
also use background bigWig files:

```{r}
bw_loci(h33_chip, bg_bwfiles = input_chip, loci = genes)
```

A name column is provided if the given `BED` file includes *loci* IDs.

You can also provide `GRanges` objects instead of `BED` files:

```{r}
genes_granges <- rtracklayer::import(genes, format = "BED")

bw_loci(h33_chip, bg_bwfiles = input_chip, loci = genes_granges)
```


### Aggregated loci

It is also possible to aggregate values by some kind of category. This is 
useful for `BED` files where ID is not an individual entity but rather a 
category, for instance, ChromHMM annotations:

```{r}
chrom_ranges <- import(chromhmm)

chrom_ranges
```

It is possible to get a single value per category using `aggregate_by` parameter.

```{r}
# If aggregate_by is not provided, you get a non-aggregated GRanges
bw_loci(h33_chip,
        bg_bwfiles = input_chip,
        loci = chrom_ranges,
        aggregate_by = "mean") 
```

**Note:** In this case you will not get a `GRanges` object but a `data.frame`.

For these calculations you **always** get some kind of **average** per locus. 
However, different ways of treating the average values are available by changing
the `aggregate_by` parameter:

- `true_mean`: This calculates a mean adding up all the values included per *locus*. 
It is most meaningful when loci represent categories rather than biological
entities.
- `mean`: This calculates a mean-of-means instead, where each *locus* mean value
is calculated and the mean of this distribution of means is returned as a result.
- `median`: This calculates a median-of-means. It is useful to perform an analysis
that is more robust to outliers, but it does not work well with sparse data.

### Profile averages

You may be interested in the **shape** the signal takes across a set of *loci*.
This is calculated by `bw_profile`. This function takes each *locus* in the 
provided bed file and slices it into a fixed number of bins (this is determined
by `bin_size` parameter). Each bin is aggregated together across loci.


```{r}
profile <- bw_profile(h33_chip, loci = genes, bin_size = 100)

head(profile)
```

**Note:**. `bw_profile` returns data as long format, since it returns more than
one value per data point. You get: mean, standard error, median value per point.
Each point is represented by index. This index determines its bin.

You can also provide `upstream` and `downstream` base pairs values to include
in this calculation.

```{r}
profile <- bw_profile(h33_chip, loci = genes, bin_size = 100,
                      upstream = 500, downstream = 500)

head(profile)
```

Additionally, you can use `mode` parameter to define how these locus are aligned
before binning. Available modes are: `start`, `end`, `center` (align the loci around
these points) or `stretch` which anchors features `start` to `end` and stretches
them when length is different.

### Per-locus profiles

This shape created by `bw_profile` function is still an average. So you may be
interested on a per-locus profile value. This is what you get with `bw_heatmap`.
`bw_heatmap` returns a `matrix` where each row is a locus and each column is
a bin. The rest of parameters are analogous to `bw_profile`:

```{r}
mat <- bw_heatmap(h33_chip, loci = genes, bin_size = 1000,
                  upstream = 5000, downstream = 5000)

# Note that bw_heatmap returns a list of matrices, so it supports arrays of 
# bwfiles as input as well.
head(mat[[1]])
```

## General shared options

Each `wigglescout` function has a different set of parameters. You can check
the documentation with `?` operator: `?bw_bins`.

Here are a few that are commonly appearing in many of the functions:

- `labels`: This parameter gives name to the value column of the `data.frame` or
`GRanges` object obtained. By default the names correspond to the file names.
- `remove_top`: Exclude a certain top quantile of the returning object. This 
value must be in [0, 1] and by default is 0 (no values removed). Underneath,
threshold value is calculated by `quantile` function and used as a threshold,
where the exact value is included (values <= threshold are returned). 
If the calculated data structure has more than one column (i.e. multiple files
are provided), quantiles are calculated over `rowMeans`. This prioritizes cases
where outliers are such in more than one sample. In the case where `NA` values
are found, these are also dropped if `remove_top > 0`.
- `per_locus_stat`: Aggregating function per locus. This is an interface with
the underlying `rtracklayer` summary function. It defaults to `mean` and it is
not recommended to change it.


```{r session-info}
sessionInfo()
```


