<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Summarizing bigWig files • wigglescout</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Summarizing bigWig files">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">wigglescout</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.19.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../articles/wigglescout.html">Get started</a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/bwtools.html">Summarizing bigWig files</a></li>
    <li><a class="dropdown-item" href="../articles/plotting.html">Visualizing bigWig signal</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/cnluzon/wigglescout/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Summarizing bigWig files</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/cnluzon/wigglescout/blob/main/vignettes/bwtools.Rmd" class="external-link"><code>vignettes/bwtools.Rmd</code></a></small>
      <div class="d-none name"><code>bwtools.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="calculating-functions">Calculating functions<a class="anchor" aria-label="anchor" href="#calculating-functions"></a>
</h2>
<p>This vignette uses the bundled data to showcase these functions.
Please refer to the quick start guide for more information about the
bundled data.</p>
<p>bigWig data can be summarized genome-wide using calculation functions
that leverage <code>rtracklayer</code> for handling bigWig files (and
<code>furrr</code> to boost performance for large use cases):
<code>bw_bins</code>, <code>bw_loci</code>, <code>bw_heatmap</code> or
<code>bw_profile</code>.</p>
<div class="section level3">
<h3 id="genome-wide-analysis">Genome-wide analysis<a class="anchor" aria-label="anchor" href="#genome-wide-analysis"></a>
</h3>
<p><code>bw_bins</code> function returns a <code>GRanges</code> object
where each entry is a locus of length <code>bin_size</code> and its
score is the mean coverage.</p>
<p><strong>Note:</strong> Small bin sizes (&lt; 5000bp) take longer to
run.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Several bwfiles can be provided at once</span></span>
<span><span class="fu"><a href="../reference/bw_bins.html">bw_bins</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">h33_chip</span>, <span class="va">input_chip</span><span class="op">)</span>,</span>
<span>        bin_size <span class="op">=</span> <span class="fl">2000</span>,</span>
<span>        genome <span class="op">=</span> <span class="st">"mm9"</span>,</span>
<span>        selection <span class="op">=</span> <span class="va">locus</span><span class="op">)</span></span>
<span><span class="co">#&gt; GRanges object with 251 ranges and 2 metadata columns:</span></span>
<span><span class="co">#&gt;         seqnames              ranges strand | sample_H33_ChIP sample_Input</span></span>
<span><span class="co">#&gt;            &lt;Rle&gt;           &lt;IRanges&gt;  &lt;Rle&gt; |       &lt;numeric&gt;    &lt;numeric&gt;</span></span>
<span><span class="co">#&gt;     [1]    chr15 102598001-102600000      * |        0.548169     0.999076</span></span>
<span><span class="co">#&gt;     [2]    chr15 102600001-102602000      * |        2.382783     0.987250</span></span>
<span><span class="co">#&gt;     [3]    chr15 102602001-102604000      * |        1.872876     0.793113</span></span>
<span><span class="co">#&gt;     [4]    chr15 102604001-102606000      * |        0.640021     0.720190</span></span>
<span><span class="co">#&gt;     [5]    chr15 102606001-102608000      * |        1.035948     1.134485</span></span>
<span><span class="co">#&gt;     ...      ...                 ...    ... .             ...          ...</span></span>
<span><span class="co">#&gt;   [247]    chr15 103090001-103092000      * |         1.33303     0.961300</span></span>
<span><span class="co">#&gt;   [248]    chr15 103092001-103094000      * |         1.04270     0.977078</span></span>
<span><span class="co">#&gt;   [249]    chr15 103094001-103096000      * |         1.47173     1.212554</span></span>
<span><span class="co">#&gt;   [250]    chr15 103096001-103098000      * |         1.44952     1.117081</span></span>
<span><span class="co">#&gt;   [251]    chr15 103098001-103100000      * |         1.34690     0.749071</span></span>
<span><span class="co">#&gt;   -------</span></span>
<span><span class="co">#&gt;   seqinfo: 35 sequences from an unspecified genome; no seqlengths</span></span></code></pre></div>
<p>It is also possible to provide a list of bigWig files to be used as
background to normalize the bin values to. For instance, in the previous
case, one could want to use the input values to normalize the H3.3 ChIP
data:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/bw_bins.html">bw_bins</a></span><span class="op">(</span><span class="va">h33_chip</span>, bg_bwfiles <span class="op">=</span> <span class="va">input_chip</span>,</span>
<span>        bin_size <span class="op">=</span> <span class="fl">2000</span>,</span>
<span>        genome <span class="op">=</span> <span class="st">"mm9"</span>,</span>
<span>        selection <span class="op">=</span> <span class="va">locus</span><span class="op">)</span></span>
<span><span class="co">#&gt; GRanges object with 251 ranges and 1 metadata column:</span></span>
<span><span class="co">#&gt;         seqnames              ranges strand | sample_H33_ChIP</span></span>
<span><span class="co">#&gt;            &lt;Rle&gt;           &lt;IRanges&gt;  &lt;Rle&gt; |       &lt;numeric&gt;</span></span>
<span><span class="co">#&gt;     [1]    chr15 102598001-102600000      * |        0.548676</span></span>
<span><span class="co">#&gt;     [2]    chr15 102600001-102602000      * |        2.413555</span></span>
<span><span class="co">#&gt;     [3]    chr15 102602001-102604000      * |        2.361423</span></span>
<span><span class="co">#&gt;     [4]    chr15 102604001-102606000      * |        0.888684</span></span>
<span><span class="co">#&gt;     [5]    chr15 102606001-102608000      * |        0.913143</span></span>
<span><span class="co">#&gt;     ...      ...                 ...    ... .             ...</span></span>
<span><span class="co">#&gt;   [247]    chr15 103090001-103092000      * |         1.38670</span></span>
<span><span class="co">#&gt;   [248]    chr15 103092001-103094000      * |         1.06716</span></span>
<span><span class="co">#&gt;   [249]    chr15 103094001-103096000      * |         1.21375</span></span>
<span><span class="co">#&gt;   [250]    chr15 103096001-103098000      * |         1.29760</span></span>
<span><span class="co">#&gt;   [251]    chr15 103098001-103100000      * |         1.79810</span></span>
<span><span class="co">#&gt;   -------</span></span>
<span><span class="co">#&gt;   seqinfo: 35 sequences from an unspecified genome; no seqlengths</span></span></code></pre></div>
<p>If you want to get the log2 ratio sample / input, select
<code>norm_mode = "log2fc"</code>:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/bw_bins.html">bw_bins</a></span><span class="op">(</span><span class="va">h33_chip</span>, bg_bwfiles <span class="op">=</span> <span class="va">input_chip</span>,</span>
<span>        bin_size <span class="op">=</span> <span class="fl">2000</span>,</span>
<span>        genome <span class="op">=</span> <span class="st">"mm9"</span>,</span>
<span>        selection <span class="op">=</span> <span class="va">locus</span>,</span>
<span>        norm_mode <span class="op">=</span> <span class="st">"log2fc"</span><span class="op">)</span></span>
<span><span class="co">#&gt; GRanges object with 251 ranges and 1 metadata column:</span></span>
<span><span class="co">#&gt;         seqnames              ranges strand | sample_H33_ChIP</span></span>
<span><span class="co">#&gt;            &lt;Rle&gt;           &lt;IRanges&gt;  &lt;Rle&gt; |       &lt;numeric&gt;</span></span>
<span><span class="co">#&gt;     [1]    chr15 102598001-102600000      * |       -0.865973</span></span>
<span><span class="co">#&gt;     [2]    chr15 102600001-102602000      * |        1.271160</span></span>
<span><span class="co">#&gt;     [3]    chr15 102602001-102604000      * |        1.239656</span></span>
<span><span class="co">#&gt;     [4]    chr15 102604001-102606000      * |       -0.170258</span></span>
<span><span class="co">#&gt;     [5]    chr15 102606001-102608000      * |       -0.131087</span></span>
<span><span class="co">#&gt;     ...      ...                 ...    ... .             ...</span></span>
<span><span class="co">#&gt;   [247]    chr15 103090001-103092000      * |       0.4716506</span></span>
<span><span class="co">#&gt;   [248]    chr15 103092001-103094000      * |       0.0937752</span></span>
<span><span class="co">#&gt;   [249]    chr15 103094001-103096000      * |       0.2794686</span></span>
<span><span class="co">#&gt;   [250]    chr15 103096001-103098000      * |       0.3758455</span></span>
<span><span class="co">#&gt;   [251]    chr15 103098001-103100000      * |       0.8464701</span></span>
<span><span class="co">#&gt;   -------</span></span>
<span><span class="co">#&gt;   seqinfo: 35 sequences from an unspecified genome; no seqlengths</span></span></code></pre></div>
<p><code>norm_mode</code> is set to <code>"fc"</code> by default, which
means it shows fold change sample / input, if input (background) is
provided.</p>
<div class="section level4">
<h4 id="supported-genomes">Supported genomes<a class="anchor" aria-label="anchor" href="#supported-genomes"></a>
</h4>
<p><code>bw_bins</code> function is based on <code>tileGenome</code>
function from <code>GenomicRanges</code>, which requires information
about chromosome length. This is obtained relying on function
<code>Seqinfo</code> from <code>GenomeInfoDb</code> package. So, in
theory, any genome ID that would return a <code>Seqinfo</code> object by
doing <code>Seqinfo(genome="your_genome")</code> should work.</p>
</div>
</div>
<div class="section level3">
<h3 id="locus-based-analysis">Locus-based analysis<a class="anchor" aria-label="anchor" href="#locus-based-analysis"></a>
</h3>
<p>These functions are useful when you have some genomic annotations you
want to look at, rather than general genome-wide trends.</p>
<div class="section level4">
<h4 id="individual-loci">Individual loci<a class="anchor" aria-label="anchor" href="#individual-loci"></a>
</h4>
<p>To calculate values on a <em>locus</em> specific manner you can use
<code>bw_loci</code> function. For instance, we can look at the genes at
this region:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/bw_loci.html">bw_loci</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">h33_chip</span>, <span class="va">input_chip</span><span class="op">)</span>, loci <span class="op">=</span> <span class="va">genes</span><span class="op">)</span></span>
<span><span class="co">#&gt; GRanges object with 26 ranges and 3 metadata columns:</span></span>
<span><span class="co">#&gt;        seqnames              ranges strand | sample_H33_ChIP sample_Input</span></span>
<span><span class="co">#&gt;           &lt;Rle&gt;           &lt;IRanges&gt;  &lt;Rle&gt; |       &lt;numeric&gt;    &lt;numeric&gt;</span></span>
<span><span class="co">#&gt;    [1]    chr15 102751562-102759245      + |         2.91694     0.989028</span></span>
<span><span class="co">#&gt;    [2]    chr15 102767284-102768948      + |         3.61492     1.082491</span></span>
<span><span class="co">#&gt;    [3]    chr15 102774493-102778377      - |         3.97043     1.135737</span></span>
<span><span class="co">#&gt;    [4]    chr15 102784957-102787132      + |         2.17507     1.122432</span></span>
<span><span class="co">#&gt;    [5]    chr15 102797227-102802328      + |         1.45977     1.072421</span></span>
<span><span class="co">#&gt;    ...      ...                 ...    ... .             ...          ...</span></span>
<span><span class="co">#&gt;   [22]    chr15 103078643-103082118      - |         5.51339      1.22775</span></span>
<span><span class="co">#&gt;   [23]    chr15 103078643-103083241      - |         5.49223      1.25069</span></span>
<span><span class="co">#&gt;   [24]    chr15 103078643-103083583      - |         5.45826      1.25206</span></span>
<span><span class="co">#&gt;   [25]    chr15 103078643-103085870      - |         4.87355      1.23020</span></span>
<span><span class="co">#&gt;   [26]    chr15 103078643-103088834      - |         4.42227      1.20272</span></span>
<span><span class="co">#&gt;               name</span></span>
<span><span class="co">#&gt;        &lt;character&gt;</span></span>
<span><span class="co">#&gt;    [1]      Hoxc13</span></span>
<span><span class="co">#&gt;    [2]      Hoxc12</span></span>
<span><span class="co">#&gt;    [3]      Hotair</span></span>
<span><span class="co">#&gt;    [4]      Hoxc11</span></span>
<span><span class="co">#&gt;    [5]      Hoxc10</span></span>
<span><span class="co">#&gt;    ...         ...</span></span>
<span><span class="co">#&gt;   [22]        Nfe2</span></span>
<span><span class="co">#&gt;   [23]        Nfe2</span></span>
<span><span class="co">#&gt;   [24]        Nfe2</span></span>
<span><span class="co">#&gt;   [25]        Nfe2</span></span>
<span><span class="co">#&gt;   [26]        Nfe2</span></span>
<span><span class="co">#&gt;   -------</span></span>
<span><span class="co">#&gt;   seqinfo: 1 sequence from an unspecified genome; no seqlengths</span></span></code></pre></div>
<p>Many of the options available for <code>bw_bins</code> are shared
with <code>bw_loci</code>, so you can also use background bigWig
files:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/bw_loci.html">bw_loci</a></span><span class="op">(</span><span class="va">h33_chip</span>, bg_bwfiles <span class="op">=</span> <span class="va">input_chip</span>, loci <span class="op">=</span> <span class="va">genes</span><span class="op">)</span></span>
<span><span class="co">#&gt; GRanges object with 26 ranges and 2 metadata columns:</span></span>
<span><span class="co">#&gt;        seqnames              ranges strand | sample_H33_ChIP        name</span></span>
<span><span class="co">#&gt;           &lt;Rle&gt;           &lt;IRanges&gt;  &lt;Rle&gt; |       &lt;numeric&gt; &lt;character&gt;</span></span>
<span><span class="co">#&gt;    [1]    chr15 102751562-102759245      + |         2.94930      Hoxc13</span></span>
<span><span class="co">#&gt;    [2]    chr15 102767284-102768948      + |         3.33944      Hoxc12</span></span>
<span><span class="co">#&gt;    [3]    chr15 102774493-102778377      - |         3.49591      Hotair</span></span>
<span><span class="co">#&gt;    [4]    chr15 102784957-102787132      + |         1.93782      Hoxc11</span></span>
<span><span class="co">#&gt;    [5]    chr15 102797227-102802328      + |         1.36119      Hoxc10</span></span>
<span><span class="co">#&gt;    ...      ...                 ...    ... .             ...         ...</span></span>
<span><span class="co">#&gt;   [22]    chr15 103078643-103082118      - |         4.49064        Nfe2</span></span>
<span><span class="co">#&gt;   [23]    chr15 103078643-103083241      - |         4.39137        Nfe2</span></span>
<span><span class="co">#&gt;   [24]    chr15 103078643-103083583      - |         4.35942        Nfe2</span></span>
<span><span class="co">#&gt;   [25]    chr15 103078643-103085870      - |         3.96159        Nfe2</span></span>
<span><span class="co">#&gt;   [26]    chr15 103078643-103088834      - |         3.67690        Nfe2</span></span>
<span><span class="co">#&gt;   -------</span></span>
<span><span class="co">#&gt;   seqinfo: 1 sequence from an unspecified genome; no seqlengths</span></span></code></pre></div>
<p>A name column is provided if the given <code>BED</code> file includes
<em>loci</em> IDs.</p>
<p>You can also provide <code>GRanges</code> objects instead of
<code>BED</code> files:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">genes_granges</span> <span class="op">&lt;-</span> <span class="fu">rtracklayer</span><span class="fu">::</span><span class="fu">import</span><span class="op">(</span><span class="va">genes</span>, format <span class="op">=</span> <span class="st">"BED"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="../reference/bw_loci.html">bw_loci</a></span><span class="op">(</span><span class="va">h33_chip</span>, bg_bwfiles <span class="op">=</span> <span class="va">input_chip</span>, loci <span class="op">=</span> <span class="va">genes_granges</span><span class="op">)</span></span>
<span><span class="co">#&gt; GRanges object with 26 ranges and 2 metadata columns:</span></span>
<span><span class="co">#&gt;        seqnames              ranges strand | sample_H33_ChIP        name</span></span>
<span><span class="co">#&gt;           &lt;Rle&gt;           &lt;IRanges&gt;  &lt;Rle&gt; |       &lt;numeric&gt; &lt;character&gt;</span></span>
<span><span class="co">#&gt;    [1]    chr15 102751562-102759245      + |         2.94930      Hoxc13</span></span>
<span><span class="co">#&gt;    [2]    chr15 102767284-102768948      + |         3.33944      Hoxc12</span></span>
<span><span class="co">#&gt;    [3]    chr15 102774493-102778377      - |         3.49591      Hotair</span></span>
<span><span class="co">#&gt;    [4]    chr15 102784957-102787132      + |         1.93782      Hoxc11</span></span>
<span><span class="co">#&gt;    [5]    chr15 102797227-102802328      + |         1.36119      Hoxc10</span></span>
<span><span class="co">#&gt;    ...      ...                 ...    ... .             ...         ...</span></span>
<span><span class="co">#&gt;   [22]    chr15 103078643-103082118      - |         4.49064        Nfe2</span></span>
<span><span class="co">#&gt;   [23]    chr15 103078643-103083241      - |         4.39137        Nfe2</span></span>
<span><span class="co">#&gt;   [24]    chr15 103078643-103083583      - |         4.35942        Nfe2</span></span>
<span><span class="co">#&gt;   [25]    chr15 103078643-103085870      - |         3.96159        Nfe2</span></span>
<span><span class="co">#&gt;   [26]    chr15 103078643-103088834      - |         3.67690        Nfe2</span></span>
<span><span class="co">#&gt;   -------</span></span>
<span><span class="co">#&gt;   seqinfo: 1 sequence from an unspecified genome; no seqlengths</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="aggregated-loci">Aggregated loci<a class="anchor" aria-label="anchor" href="#aggregated-loci"></a>
</h4>
<p>It is also possible to aggregate values by some kind of category.
This is useful for <code>BED</code> files where ID is not an individual
entity but rather a category, for instance, ChromHMM annotations:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">chrom_ranges</span> <span class="op">&lt;-</span> <span class="fu">import</span><span class="op">(</span><span class="va">chromhmm</span><span class="op">)</span></span>
<span></span>
<span><span class="va">chrom_ranges</span></span>
<span><span class="co">#&gt; GRanges object with 164 ranges and 2 metadata columns:</span></span>
<span><span class="co">#&gt;         seqnames              ranges strand |               name     score</span></span>
<span><span class="co">#&gt;            &lt;Rle&gt;           &lt;IRanges&gt;  &lt;Rle&gt; |        &lt;character&gt; &lt;numeric&gt;</span></span>
<span><span class="co">#&gt;     [1]    chr15 102602601-102606600      * |     12_Heterochrom         0</span></span>
<span><span class="co">#&gt;     [2]    chr15 102606601-102607600      * |       11_Repressed         0</span></span>
<span><span class="co">#&gt;     [3]    chr15 102615401-102615800      * |       15_Insulator         0</span></span>
<span><span class="co">#&gt;     [4]    chr15 102615801-102616800      * |       11_Repressed         0</span></span>
<span><span class="co">#&gt;     [5]    chr15 102616801-102626600      * |     12_Heterochrom         0</span></span>
<span><span class="co">#&gt;     ...      ...                 ...    ... .                ...       ...</span></span>
<span><span class="co">#&gt;   [160]    chr15 103086201-103088200      * |         2_Weak_Txn         0</span></span>
<span><span class="co">#&gt;   [161]    chr15 102627601-102636000      * |     12_Heterochrom         0</span></span>
<span><span class="co">#&gt;   [162]    chr15 102891201-102891600      * | 10_Poised_Promoter         0</span></span>
<span><span class="co">#&gt;   [163]    chr15 103017601-103040000      * |   1_Txn_Elongation         0</span></span>
<span><span class="co">#&gt;   [164]    chr15 102758201-102761800      * |       11_Repressed         0</span></span>
<span><span class="co">#&gt;   -------</span></span>
<span><span class="co">#&gt;   seqinfo: 1 sequence from an unspecified genome; no seqlengths</span></span></code></pre></div>
<p>It is possible to get a single value per category using
<code>aggregate_by</code> parameter.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># If aggregate_by is not provided, you get a non-aggregated GRanges</span></span>
<span><span class="fu"><a href="../reference/bw_loci.html">bw_loci</a></span><span class="op">(</span><span class="va">h33_chip</span>,</span>
<span>        bg_bwfiles <span class="op">=</span> <span class="va">input_chip</span>,</span>
<span>        loci <span class="op">=</span> <span class="va">chrom_ranges</span>,</span>
<span>        aggregate_by <span class="op">=</span> <span class="st">"mean"</span><span class="op">)</span> </span>
<span><span class="co">#&gt;                    sample_H33_ChIP</span></span>
<span><span class="co">#&gt; 1_Txn_Elongation          4.089008</span></span>
<span><span class="co">#&gt; 2_Weak_Txn                2.438608</span></span>
<span><span class="co">#&gt; 4_Poised_Enhancer         3.298239</span></span>
<span><span class="co">#&gt; 5_Active_Promoter         3.807038</span></span>
<span><span class="co">#&gt; 6_Strong_Enhancer         4.691568</span></span>
<span><span class="co">#&gt; 7_Active_Promoter         5.561953</span></span>
<span><span class="co">#&gt; 8_Strong_Enhancer         6.381945</span></span>
<span><span class="co">#&gt; 9_Strong_Enhancer         3.320948</span></span>
<span><span class="co">#&gt; 9_Txn_Transition          3.458106</span></span>
<span><span class="co">#&gt; 10_Poised_Promoter        2.863107</span></span>
<span><span class="co">#&gt; 11_Repressed              1.961251</span></span>
<span><span class="co">#&gt; 12_Heterochrom            1.038954</span></span>
<span><span class="co">#&gt; 14_Heterochrom            3.597231</span></span>
<span><span class="co">#&gt; 15_Insulator              2.400263</span></span></code></pre></div>
<p><strong>Note:</strong> In this case you will not get a
<code>GRanges</code> object but a <code>data.frame</code>.</p>
<p>For these calculations you <strong>always</strong> get some kind of
<strong>average</strong> per locus. However, different ways of treating
the average values are available by changing the
<code>aggregate_by</code> parameter:</p>
<ul>
<li>
<code>true_mean</code>: This calculates a mean adding up all the
values included per <em>locus</em>. It is most meaningful when loci
represent categories rather than biological entities.</li>
<li>
<code>mean</code>: This calculates a mean-of-means instead, where
each <em>locus</em> mean value is calculated and the mean of this
distribution of means is returned as a result.</li>
<li>
<code>median</code>: This calculates a median-of-means. It is useful
to perform an analysis that is more robust to outliers, but it does not
work well with sparse data.</li>
</ul>
</div>
<div class="section level4">
<h4 id="profile-averages">Profile averages<a class="anchor" aria-label="anchor" href="#profile-averages"></a>
</h4>
<p>You may be interested in the <strong>shape</strong> the signal takes
across a set of <em>loci</em>. This is calculated by
<code>bw_profile</code>. This function takes each <em>locus</em> in the
provided bed file and slices it into a fixed number of bins (this is
determined by <code>bin_size</code> parameter). Each bin is aggregated
together across loci.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">profile</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/bw_profile.html">bw_profile</a></span><span class="op">(</span><span class="va">h33_chip</span>, loci <span class="op">=</span> <span class="va">genes</span>, bin_size <span class="op">=</span> <span class="fl">100</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">profile</span><span class="op">)</span></span>
<span><span class="co">#&gt;       mean   sderror   median index          sample</span></span>
<span><span class="co">#&gt; 1 2.651766 0.4211416 1.868056     1 sample_H33_ChIP</span></span>
<span><span class="co">#&gt; 2 2.353699 0.3952149 1.744718     2 sample_H33_ChIP</span></span>
<span><span class="co">#&gt; 3 2.372053 0.4093088 1.675340     3 sample_H33_ChIP</span></span>
<span><span class="co">#&gt; 4 2.516681 0.2644566 2.112161     4 sample_H33_ChIP</span></span>
<span><span class="co">#&gt; 5 2.288359 0.2839340 1.683047     5 sample_H33_ChIP</span></span>
<span><span class="co">#&gt; 6 2.415918 0.3109912 1.837221     6 sample_H33_ChIP</span></span></code></pre></div>
<p><strong>Note:</strong>. <code>bw_profile</code> returns data as long
format, since it returns more than one value per data point. You get:
mean, standard error, median value per point. Each point is represented
by index. This index determines its bin.</p>
<p>You can also provide <code>upstream</code> and
<code>downstream</code> base pairs values to include in this
calculation.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">profile</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/bw_profile.html">bw_profile</a></span><span class="op">(</span><span class="va">h33_chip</span>, loci <span class="op">=</span> <span class="va">genes</span>, bin_size <span class="op">=</span> <span class="fl">100</span>,</span>
<span>                      upstream <span class="op">=</span> <span class="fl">500</span>, downstream <span class="op">=</span> <span class="fl">500</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">profile</span><span class="op">)</span></span>
<span><span class="co">#&gt;       mean   sderror   median index          sample</span></span>
<span><span class="co">#&gt; 1 3.428869 0.5597654 2.405090     1 sample_H33_ChIP</span></span>
<span><span class="co">#&gt; 2 3.949016 0.6592606 2.949832     2 sample_H33_ChIP</span></span>
<span><span class="co">#&gt; 3 3.247899 0.4771775 2.590096     3 sample_H33_ChIP</span></span>
<span><span class="co">#&gt; 4 3.229546 0.4894977 2.058201     4 sample_H33_ChIP</span></span>
<span><span class="co">#&gt; 5 3.079227 0.4333806 2.497594     5 sample_H33_ChIP</span></span>
<span><span class="co">#&gt; 6 3.639487 0.3295900 3.067325     6 sample_H33_ChIP</span></span></code></pre></div>
<p>Additionally, you can use <code>mode</code> parameter to define how
these locus are aligned before binning. Available modes are:
<code>start</code>, <code>end</code>, <code>center</code> (align the
loci around these points) or <code>stretch</code> which anchors features
<code>start</code> to <code>end</code> and stretches them when length is
different.</p>
</div>
<div class="section level4">
<h4 id="per-locus-profiles">Per-locus profiles<a class="anchor" aria-label="anchor" href="#per-locus-profiles"></a>
</h4>
<p>This shape created by <code>bw_profile</code> function is still an
average. So you may be interested on a per-locus profile value. This is
what you get with <code>bw_heatmap</code>. <code>bw_heatmap</code>
returns a <code>matrix</code> where each row is a locus and each column
is a bin. The rest of parameters are analogous to
<code>bw_profile</code>:</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mat</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/bw_heatmap.html">bw_heatmap</a></span><span class="op">(</span><span class="va">h33_chip</span>, loci <span class="op">=</span> <span class="va">genes</span>, bin_size <span class="op">=</span> <span class="fl">1000</span>,</span>
<span>                  upstream <span class="op">=</span> <span class="fl">5000</span>, downstream <span class="op">=</span> <span class="fl">5000</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Note that bw_heatmap returns a list of matrices, so it supports arrays of </span></span>
<span><span class="co"># bwfiles as input as well.</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">mat</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="co">#&gt;          [,1]      [,2]      [,3]      [,4]     [,5]     [,6]     [,7]</span></span>
<span><span class="co">#&gt; [1,] 1.284604 1.3293102 1.7112652 2.8632013 3.007691 2.800395 1.766877</span></span>
<span><span class="co">#&gt; [2,] 4.952701 6.5925578 6.4507772 5.3514131 3.370813 2.863804 3.339249</span></span>
<span><span class="co">#&gt; [3,] 1.817719 1.8833939 2.0797163 1.5973820 3.622250 5.756086 3.083041</span></span>
<span><span class="co">#&gt; [4,] 1.666320 2.4688960 1.5384674 1.5898725 1.831000 2.566131 1.285138</span></span>
<span><span class="co">#&gt; [5,] 1.640442 2.2717496 1.9870765 1.9248061 2.705879 2.634167 1.318919</span></span>
<span><span class="co">#&gt; [6,] 1.259195 0.9254259 0.8669583 0.9295532 2.033445 2.275877 1.419363</span></span>
<span><span class="co">#&gt;           [,8]      [,9]    [,10]    [,11]    [,12]    [,13]    [,14]</span></span>
<span><span class="co">#&gt; [1,] 3.3833293 4.3805848 1.667400 1.561681 3.661995 4.761350 6.651459</span></span>
<span><span class="co">#&gt; [2,] 3.0416741 5.1382823 5.887521 9.196965 4.561860 3.057636 3.688586</span></span>
<span><span class="co">#&gt; [3,] 2.7693828 4.2945841 4.607247 3.440608 3.096550 7.539874 8.274333</span></span>
<span><span class="co">#&gt; [4,] 2.6849959 2.1768526 2.917202 1.981451 1.685489 5.213727 2.688018</span></span>
<span><span class="co">#&gt; [5,] 0.9578958 0.7213599 1.690166 2.632981 2.310449 1.297612 2.311900</span></span>
<span><span class="co">#&gt; [6,] 2.3248209 3.5973599 3.592973 1.099763 1.261885 4.124952 3.420550</span></span></code></pre></div>
</div>
</div>
<div class="section level3">
<h3 id="general-shared-options">General shared options<a class="anchor" aria-label="anchor" href="#general-shared-options"></a>
</h3>
<p>Each <code>wigglescout</code> function has a different set of
parameters. You can check the documentation with <code>?</code>
operator: <code><a href="../reference/bw_bins.html">?bw_bins</a></code>.</p>
<p>Here are a few that are commonly appearing in many of the
functions:</p>
<ul>
<li>
<code>labels</code>: This parameter gives name to the value column
of the <code>data.frame</code> or <code>GRanges</code> object obtained.
By default the names correspond to the file names.</li>
<li>
<code>remove_top</code>: Exclude a certain top quantile of the
returning object. This value must be in [0, 1] and by default is 0 (no
values removed). Underneath, threshold value is calculated by
<code>quantile</code> function and used as a threshold, where the exact
value is included (values &lt;= threshold are returned). If the
calculated data structure has more than one column (i.e. multiple files
are provided), quantiles are calculated over <code>rowMeans</code>. This
prioritizes cases where outliers are such in more than one sample. In
the case where <code>NA</code> values are found, these are also dropped
if <code>remove_top &gt; 0</code>.</li>
<li>
<code>per_locus_stat</code>: Aggregating function per locus. This is
an interface with the underlying <code>rtracklayer</code> summary
function. It defaults to <code>mean</code> and it is not recommended to
change it.</li>
</ul>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; R version 4.4.1 (2024-06-14)</span></span>
<span><span class="co">#&gt; Platform: x86_64-pc-linux-gnu</span></span>
<span><span class="co">#&gt; Running under: Ubuntu 22.04.5 LTS</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">#&gt; LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so;  LAPACK version 3.10.0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt;  [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8       </span></span>
<span><span class="co">#&gt;  [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8   </span></span>
<span><span class="co">#&gt;  [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C          </span></span>
<span><span class="co">#&gt; [10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C   </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; time zone: UTC</span></span>
<span><span class="co">#&gt; tzcode source: system (glibc)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] stats4    stats     graphics  grDevices utils     datasets  methods  </span></span>
<span><span class="co">#&gt; [8] base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt; [1] purrr_1.0.2          wigglescout_0.19.0   rtracklayer_1.64.0  </span></span>
<span><span class="co">#&gt; [4] GenomicRanges_1.56.1 GenomeInfoDb_1.40.1  IRanges_2.38.1      </span></span>
<span><span class="co">#&gt; [7] S4Vectors_0.42.1     BiocGenerics_0.50.0  ggplot2_3.5.1       </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;  [1] SummarizedExperiment_1.34.0 beeswarm_0.4.0             </span></span>
<span><span class="co">#&gt;  [3] gtable_0.3.5                rjson_0.2.23               </span></span>
<span><span class="co">#&gt;  [5] xfun_0.48                   bslib_0.8.0                </span></span>
<span><span class="co">#&gt;  [7] lattice_0.22-6              Biobase_2.64.0             </span></span>
<span><span class="co">#&gt;  [9] vctrs_0.6.5                 tools_4.4.1                </span></span>
<span><span class="co">#&gt; [11] bitops_1.0-9                generics_0.1.3             </span></span>
<span><span class="co">#&gt; [13] curl_5.2.3                  parallel_4.4.1             </span></span>
<span><span class="co">#&gt; [15] tibble_3.2.1                fansi_1.0.6                </span></span>
<span><span class="co">#&gt; [17] pkgconfig_2.0.3             Matrix_1.7-0               </span></span>
<span><span class="co">#&gt; [19] RColorBrewer_1.1-3          desc_1.4.3                 </span></span>
<span><span class="co">#&gt; [21] lifecycle_1.0.4             GenomeInfoDbData_1.2.12    </span></span>
<span><span class="co">#&gt; [23] stringr_1.5.1               compiler_4.4.1             </span></span>
<span><span class="co">#&gt; [25] Rsamtools_2.20.0            textshaping_0.4.0          </span></span>
<span><span class="co">#&gt; [27] Biostrings_2.72.1           munsell_0.5.1              </span></span>
<span><span class="co">#&gt; [29] codetools_0.2-20            vipor_0.4.7                </span></span>
<span><span class="co">#&gt; [31] htmltools_0.5.8.1           sass_0.4.9                 </span></span>
<span><span class="co">#&gt; [33] RCurl_1.98-1.16             yaml_2.3.10                </span></span>
<span><span class="co">#&gt; [35] tidyr_1.3.1                 furrr_0.3.1                </span></span>
<span><span class="co">#&gt; [37] pillar_1.9.0                pkgdown_2.1.1              </span></span>
<span><span class="co">#&gt; [39] crayon_1.5.3                jquerylib_0.1.4            </span></span>
<span><span class="co">#&gt; [41] BiocParallel_1.38.0         DelayedArray_0.30.1        </span></span>
<span><span class="co">#&gt; [43] cachem_1.1.0                abind_1.4-8                </span></span>
<span><span class="co">#&gt; [45] parallelly_1.38.0           tidyselect_1.2.1           </span></span>
<span><span class="co">#&gt; [47] digest_0.6.37               stringi_1.8.4              </span></span>
<span><span class="co">#&gt; [49] future_1.34.0               listenv_0.9.1              </span></span>
<span><span class="co">#&gt; [51] dplyr_1.1.4                 restfulr_0.0.15            </span></span>
<span><span class="co">#&gt; [53] fastmap_1.2.0               grid_4.4.1                 </span></span>
<span><span class="co">#&gt; [55] SparseArray_1.4.8           colorspace_2.1-1           </span></span>
<span><span class="co">#&gt; [57] cli_3.6.3                   magrittr_2.0.3             </span></span>
<span><span class="co">#&gt; [59] S4Arrays_1.4.1              XML_3.99-0.17              </span></span>
<span><span class="co">#&gt; [61] utf8_1.2.4                  withr_3.0.1                </span></span>
<span><span class="co">#&gt; [63] scales_1.3.0                UCSC.utils_1.0.0           </span></span>
<span><span class="co">#&gt; [65] ggbeeswarm_0.7.2            rmarkdown_2.28             </span></span>
<span><span class="co">#&gt; [67] XVector_0.44.0              httr_1.4.7                 </span></span>
<span><span class="co">#&gt; [69] globals_0.16.3              matrixStats_1.4.1          </span></span>
<span><span class="co">#&gt; [71] ragg_1.3.3                  evaluate_1.0.0             </span></span>
<span><span class="co">#&gt; [73] ggrastr_1.0.2               knitr_1.48                 </span></span>
<span><span class="co">#&gt; [75] BiocIO_1.14.0               rlang_1.1.4                </span></span>
<span><span class="co">#&gt; [77] glue_1.8.0                  jsonlite_1.8.9             </span></span>
<span><span class="co">#&gt; [79] R6_2.5.1                    MatrixGenerics_1.16.0      </span></span>
<span><span class="co">#&gt; [81] GenomicAlignments_1.40.0    systemfonts_1.1.0          </span></span>
<span><span class="co">#&gt; [83] fs_1.6.4                    zlibbioc_1.50.0</span></span></code></pre></div>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Carmen Navarro.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>





  </body>
</html>
