<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Visualizing bigWig signal • wigglescout</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Visualizing bigWig signal">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">wigglescout</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.19.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../articles/wigglescout.html">Get started</a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/bwtools.html">Summarizing bigWig files</a></li>
    <li><a class="dropdown-item" href="../articles/plotting.html">Visualizing bigWig signal</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/cnluzon/wigglescout/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Visualizing bigWig signal</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/cnluzon/wigglescout/blob/main/vignettes/plotting.Rmd" class="external-link"><code>vignettes/plotting.Rmd</code></a></small>
      <div class="d-none name"><code>plotting.Rmd</code></div>
    </div>

    
    
<p><code>wigglescout</code> provides a set of out-of-the-box functions
to visualize bigWig values genome-wide or on a defined set of loci:</p>
<ul>
<li>
<code>plot_bw_bins_scatter</code>. A genome-wide bin-based
scatterplot. Needs <code>genome</code> and <code>bin_size</code>
parameters.</li>
<li>
<code>plot_bw_loci_scatter</code>. A locus-based scatterplot. Needs
a <code>loci</code> file or <code>GRanges</code> object.</li>
<li>
<code>plot_bw_bins_violin</code>. A genome-wide bin-based
violinplot. Needs <code>genome</code> and <code>bin_size</code>
parameters.</li>
<li>
<code>plot_bw_loci_summary_heatmap</code>. Summarized heatmap over a
categorized BED file. This works great with annotations like ChromHMM,
but other BED files can be used, as long as their name field is
meaningful (provides a category). An example could be a
<code>genes.bed</code> file where genes are labelled according to
expression, low, med, high.</li>
<li>
<code>plot_bw_profile</code>. A profile plot over a set of
loci.</li>
<li>
<code>plot_bw_heatmap</code>. A per-locus heatmap.</li>
</ul>
<p>As you may have noticed, <code>loci</code> functions take as
parameter a <code>loci</code> value, which can be a BED file or a
<code>GRanges</code> object. If you provide a <code>GRanges</code>
object, chances are that you need to provide a <code>label</code>, as
there is no way to infer this from the <code>GRanges</code> object. On
the other hand, whenever bigWig or BED files are provided, they are
labeled by default with their names.</p>
<p>You can find more detailed information about these below.</p>
<div class="section level2">
<h2 id="genome-wide-plots">Genome-wide plots<a class="anchor" aria-label="anchor" href="#genome-wide-plots"></a>
</h2>
<p>Bin-based genome-wide analysis is currently supported in two
formats:</p>
<ul>
<li>
<code>plot_bw_bins_violin</code>. Makes a violin plot of one or more
bigWig files.</li>
<li>
<code>plot_bw_bins_scatter</code>. Makes a scatter plot of two
different bigWig files against each other.</li>
</ul>
<p>Both plot types allow for background bigWig signal to be provided
(i.e. input).</p>
<div class="section level3">
<h3 id="violin-plots">Violin plots<a class="anchor" aria-label="anchor" href="#violin-plots"></a>
</h3>
<p>For a set of bigWig files, it is possible to visualize the
genome-wide signal distribution. For example:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_bw_bins_violin.html">plot_bw_bins_violin</a></span><span class="op">(</span><span class="va">h33_chip</span>, bin_size <span class="op">=</span> <span class="fl">2000</span>, selection <span class="op">=</span> <span class="va">locus</span><span class="op">)</span></span></code></pre></div>
<p><img src="plotting_files/figure-html/unnamed-chunk-2-1.png" width="672"></p>
<p>Note that <code>bin_size</code> parameter influences the time any of
these functions take to run. Reasonable resolution for this is 5000 or
10000.</p>
<p>It is possible to plot several bigWig files at a time:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_bw_bins_violin.html">plot_bw_bins_violin</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">h33_chip</span>, <span class="va">input_chip</span><span class="op">)</span>, bin_size <span class="op">=</span> <span class="fl">2000</span>, selection <span class="op">=</span> <span class="va">locus</span><span class="op">)</span></span></code></pre></div>
<p><img src="plotting_files/figure-html/unnamed-chunk-3-1.png" width="672"></p>
<p>Additionally, one can highlight bins that overlap with a certain set
of interest loci. This is done via the <code>highlight</code>
parameter:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_bw_bins_violin.html">plot_bw_bins_violin</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">h33_chip</span>, <span class="va">input_chip</span><span class="op">)</span>,</span>
<span>  bin_size <span class="op">=</span> <span class="fl">2000</span>,</span>
<span>  highlight <span class="op">=</span> <span class="va">genes</span>,</span>
<span>  highlight_color <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"red"</span>, <span class="st">"red"</span><span class="op">)</span>, <span class="co"># It is possible to provide different colors</span></span>
<span>  selection <span class="op">=</span> <span class="va">locus</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="plotting_files/figure-html/unnamed-chunk-4-1.png" width="672"></p>
<p>It is also possible to provide a list of bigWig files to be used as
background to normalize the bin values to. For instance, in the previous
case, one could want to use the input values to normalize the H3.3 ChIP
data:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># bg_bwfiles parameter is set now to input_chip</span></span>
<span><span class="fu"><a href="../reference/plot_bw_bins_violin.html">plot_bw_bins_violin</a></span><span class="op">(</span></span>
<span>  <span class="va">h33_chip</span>, </span>
<span>  bg_bwfiles <span class="op">=</span> <span class="va">input_chip</span>,</span>
<span>  bin_size <span class="op">=</span> <span class="fl">2000</span>,</span>
<span>  highlight <span class="op">=</span> <span class="va">genes</span>,</span>
<span>  highlight_color <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"red"</span><span class="op">)</span>,</span>
<span>  selection <span class="op">=</span> <span class="va">locus</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="plotting_files/figure-html/unnamed-chunk-5-1.png" width="672"></p>
<p>If <code>norm_mode</code> is provided, it is possible to
log-transform these values as well:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># bg_bwfiles parameter is set now to input_chip</span></span>
<span><span class="fu"><a href="../reference/plot_bw_bins_violin.html">plot_bw_bins_violin</a></span><span class="op">(</span></span>
<span>  <span class="va">h33_chip</span>, </span>
<span>  bg_bwfiles <span class="op">=</span> <span class="va">input_chip</span>,</span>
<span>  bin_size <span class="op">=</span> <span class="fl">2000</span>,</span>
<span>  highlight <span class="op">=</span> <span class="va">genes</span>,</span>
<span>  highlight_color <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"#000088"</span><span class="op">)</span>, <span class="co"># Colors can be also HTML codes</span></span>
<span>  norm_mode <span class="op">=</span> <span class="st">"log2fc"</span>,</span>
<span>  selection <span class="op">=</span> <span class="va">locus</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="plotting_files/figure-html/unnamed-chunk-6-1.png" width="672"></p>
</div>
<div class="section level3">
<h3 id="genome-wide-scatterplots">Genome-wide scatterplots<a class="anchor" aria-label="anchor" href="#genome-wide-scatterplots"></a>
</h3>
<p>You can compare two signals against each other on a scatter plot. To
do so, you can use <code>plot_bw_bins_scatter</code>. This function
works very similarly to <code>plot_bw_bins_violin</code> but instead of
accepting an array of bigWig files, it accepts separate <code>x</code>
and <code>y</code> parameters and the corresponding optional
<code>bg_x</code> and <code>bg_y</code>, <code>norm_mode_x</code>,
<code>norm_mode_y</code>. You can see more details in the function
documentation.</p>
<p>One example would be looking at H3.3 versus H3K9me3 in this sample
data:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_bw_bins_scatter.html">plot_bw_bins_scatter</a></span><span class="op">(</span></span>
<span>  x <span class="op">=</span> <span class="va">h33_chip</span>,</span>
<span>  y <span class="op">=</span> <span class="va">h3k9me3_chip</span>,</span>
<span>  bin_size <span class="op">=</span> <span class="fl">2000</span>,</span>
<span>  selection <span class="op">=</span> <span class="va">locus</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="plotting_files/figure-html/unnamed-chunk-7-1.png" width="672"></p>
<p>This is useful to see relationships between marks. In this case, bins
where H3.3 signal is high, H3K9me3 signal is low and viceversa.</p>
<p>The same way as with <code>plot_bw_bins_violin</code>, it is possible
to specify a set of loci to highlight within the bins, and normalize to
given signals:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_bw_bins_scatter.html">plot_bw_bins_scatter</a></span><span class="op">(</span></span>
<span>  x <span class="op">=</span> <span class="va">h33_chip</span>,</span>
<span>  bg_x <span class="op">=</span> <span class="va">input_chip</span>,</span>
<span>  y <span class="op">=</span> <span class="va">h3k9me3_chip</span>,</span>
<span>  bg_y <span class="op">=</span> <span class="va">input_chip</span>,</span>
<span>  bin_size <span class="op">=</span> <span class="fl">2000</span>,</span>
<span>  highlight <span class="op">=</span> <span class="va">genes</span>,</span>
<span>  highlight_colors <span class="op">=</span> <span class="st">"red"</span>,</span>
<span>  highlight_label <span class="op">=</span> <span class="st">"genes"</span>, <span class="co"># It is also possible to label the groups</span></span>
<span>  norm_mode_x <span class="op">=</span> <span class="st">"log2fc"</span>, <span class="co"># Each norm mode is provided separately</span></span>
<span>  selection <span class="op">=</span> <span class="va">locus</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="plotting_files/figure-html/unnamed-chunk-8-1.png" width="672"></p>
</div>
</div>
<div class="section level2">
<h2 id="locus-specific-plots">Locus-specific plots<a class="anchor" aria-label="anchor" href="#locus-specific-plots"></a>
</h2>
<p>It is also possible to look at the ChIP data in a more specific way.
For this, <code>elsasserlib</code> provides the following functions:</p>
<ul>
<li>
<code>plot_bw_bed_summary_heatmap</code>: Summarizes the values
across a BED file, using name field as category.</li>
<li>
<code>plot_bw_profile</code>: Creates a profile plot across a set of
loci.</li>
<li>
<code>plot_bw_heatmap</code>: Creates a heatmap across a set of
loci.</li>
</ul>
<div class="section level3">
<h3 id="summary-heatmaps">Summary heatmaps<a class="anchor" aria-label="anchor" href="#summary-heatmaps"></a>
</h3>
<p>These plots are meant to provide a genome-wide overview of the signal
across different biologically-relevant categories. They get a list of
bigWig files and a BED file as input. Each bigWig file will represent a
row in the heatmap, and each cell will be an aggregate value of the
corresponding bigWig file in a given category in the BED file, for
example:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># ChromHMM is a genome-wide annotation according to epigenetics marks. Each</span></span>
<span><span class="co"># locus is tagged by a category. And the amount of categories must be limited.</span></span>
<span><span class="co"># In this case, it is fifteen. </span></span>
<span><span class="va">chrom_values</span> <span class="op">&lt;-</span> <span class="fu">import</span><span class="op">(</span><span class="va">chromhmm</span>, format <span class="op">=</span> <span class="st">"BED"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">chrom_values</span><span class="op">)</span></span>
<span><span class="co">#&gt; GRanges object with 6 ranges and 2 metadata columns:</span></span>
<span><span class="co">#&gt;       seqnames              ranges strand |           name     score</span></span>
<span><span class="co">#&gt;          &lt;Rle&gt;           &lt;IRanges&gt;  &lt;Rle&gt; |    &lt;character&gt; &lt;numeric&gt;</span></span>
<span><span class="co">#&gt;   [1]    chr15 102602601-102606600      * | 12_Heterochrom         0</span></span>
<span><span class="co">#&gt;   [2]    chr15 102606601-102607600      * |   11_Repressed         0</span></span>
<span><span class="co">#&gt;   [3]    chr15 102615401-102615800      * |   15_Insulator         0</span></span>
<span><span class="co">#&gt;   [4]    chr15 102615801-102616800      * |   11_Repressed         0</span></span>
<span><span class="co">#&gt;   [5]    chr15 102616801-102626600      * | 12_Heterochrom         0</span></span>
<span><span class="co">#&gt;   [6]    chr15 102626601-102627600      * |   11_Repressed         0</span></span>
<span><span class="co">#&gt;   -------</span></span>
<span><span class="co">#&gt;   seqinfo: 1 sequence from an unspecified genome; no seqlengths</span></span>
<span></span>
<span><span class="fu"><a href="../reference/plot_bw_loci_summary_heatmap.html">plot_bw_loci_summary_heatmap</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">h33_chip</span>, <span class="va">h3k9me3_chip</span><span class="op">)</span>,</span>
<span>  loci <span class="op">=</span> <span class="va">chromhmm</span>,</span>
<span>  labels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"H33"</span>, <span class="st">"H3K9me3"</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="plotting_files/figure-html/unnamed-chunk-9-1.png" width="672"></p>
<p>Since this example is reduced to a set of HOXC genes, you can see how
H3.3 signal is high and H3K9me3 is generally low.</p>
<p>It is possible as in the previous types of plots to provide bigWig
files as normalization / input tracks and use them in the same way as
before:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_bw_loci_summary_heatmap.html">plot_bw_loci_summary_heatmap</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">h33_chip</span>, <span class="va">h3k9me3_chip</span><span class="op">)</span>,</span>
<span>  bg_bwfiles <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">input_chip</span>, <span class="va">input_chip</span><span class="op">)</span>,</span>
<span>  loci <span class="op">=</span> <span class="va">chromhmm</span>,</span>
<span>  labels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"H33"</span>, <span class="st">"H3K9me3"</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="plotting_files/figure-html/unnamed-chunk-10-1.png" width="672"></p>
<div class="section level4">
<h4 id="how-the-values-are-aggregated">How the values are aggregated<a class="anchor" aria-label="anchor" href="#how-the-values-are-aggregated"></a>
</h4>
<p>Since this function is aggregating the values, it is possible to
specify the way the per-locus values are aggregated. Default is
<strong>true mean</strong>. What it means is that all the coverage
values are binned together and the mean is performed as if each category
is one big locus. Another possibilities are: <strong>mean of
means</strong> (<code>aggregate_by = "mean"</code>), which will
calculate mean coverage per locus and then the aggregate value will be
the mean of those mean values, and <strong>median of means</strong>
(<code>aggregate_by = "median"</code>). Median per-locus coverage values
tend to be noisy, so a median of means could be useful in some
cases.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_bw_loci_summary_heatmap.html">plot_bw_loci_summary_heatmap</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">h33_chip</span>, <span class="va">h3k9me3_chip</span><span class="op">)</span>,</span>
<span>  bg_bwfiles <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">input_chip</span>, <span class="va">input_chip</span><span class="op">)</span>,</span>
<span>  loci <span class="op">=</span> <span class="va">chromhmm</span>,,</span>
<span>  labels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"H33"</span>, <span class="st">"H3K9me3"</span><span class="op">)</span>,</span>
<span>  aggregate_by <span class="op">=</span> <span class="st">"median"</span>   <span class="co"># Note this is a string value, not a function</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="plotting_files/figure-html/unnamed-chunk-11-1.png" width="672"></p>
</div>
</div>
<div class="section level3">
<h3 id="profile-plots">Profile plots<a class="anchor" aria-label="anchor" href="#profile-plots"></a>
</h3>
<p>Another way to look at how the ChIP data looks is to summarize a
profile across a set of <em>loci</em>. In order to do that you can use
the <code>plot_bw_profile</code> function:</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_bw_profile.html">plot_bw_profile</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">h33_chip</span>, <span class="va">h3k9me3_chip</span>, <span class="va">input_chip</span><span class="op">)</span>,</span>
<span>  loci <span class="op">=</span> <span class="va">genes</span>,</span>
<span>  labels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"H3.3"</span>, <span class="st">"H3K9me3"</span>, <span class="st">"Input"</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="plotting_files/figure-html/unnamed-chunk-12-1.png" width="672"></p>
<p>The default mode is called <code>stretch</code>, where start (3’ end)
and end (5’ end) of <em>loci</em> are aligned and the length of the
<em>locus</em> is stretched to a given pseudolength (<code>middle</code>
parameter). If this parameter is not given, the pseudolength will be
assigned as the median value of the length distribution of
<em>loci</em>, in an attempt to match a realistic length.</p>
<p>It is possible to align only 3’ ends, 5’ ends or center of loci,
similarly to what formatting text alignment would do:</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_bw_profile.html">plot_bw_profile</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">h33_chip</span>, <span class="va">h3k9me3_chip</span>, <span class="va">input_chip</span><span class="op">)</span>,</span>
<span>  loci <span class="op">=</span> <span class="va">genes</span>,</span>
<span>  labels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"H3.3"</span>, <span class="st">"H3K9me3"</span>, <span class="st">"Input"</span><span class="op">)</span>,</span>
<span>  mode <span class="op">=</span> <span class="st">"start"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="plotting_files/figure-html/unnamed-chunk-13-1.png" width="672"></p>
<p>It is also possible to specify the upstream and downstream
lengths:</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_bw_profile.html">plot_bw_profile</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">h33_chip</span>, <span class="va">h3k9me3_chip</span>, <span class="va">input_chip</span><span class="op">)</span>,</span>
<span>  loci <span class="op">=</span> <span class="va">genes</span>,</span>
<span>  labels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"H3.3"</span>, <span class="st">"H3K9me3"</span>, <span class="st">"Input"</span><span class="op">)</span>,</span>
<span>  mode <span class="op">=</span> <span class="st">"center"</span>,</span>
<span>  upstream <span class="op">=</span> <span class="fl">3000</span>,</span>
<span>  downstream <span class="op">=</span> <span class="fl">3000</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="plotting_files/figure-html/unnamed-chunk-14-1.png" width="672"></p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_bw_profile.html">plot_bw_profile</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">h33_chip</span>, <span class="va">h3k9me3_chip</span>, <span class="va">input_chip</span><span class="op">)</span>,</span>
<span>  loci <span class="op">=</span> <span class="va">genes</span>,</span>
<span>  labels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"H3.3"</span>, <span class="st">"H3K9me3"</span>, <span class="st">"Input"</span><span class="op">)</span>,</span>
<span>  mode <span class="op">=</span> <span class="st">"end"</span>,</span>
<span>  upstream <span class="op">=</span> <span class="fl">7000</span>,</span>
<span>  downstream <span class="op">=</span> <span class="fl">3000</span>,</span>
<span>  show_error <span class="op">=</span> <span class="cn">TRUE</span> <span class="co"># Turn error estimates on</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="plotting_files/figure-html/unnamed-chunk-15-1.png" width="672"></p>
</div>
<div class="section level3">
<h3 id="heatmaps">Heatmaps<a class="anchor" aria-label="anchor" href="#heatmaps"></a>
</h3>
<p>Finally, it is possible to look at the profile across a set of
<em>loci</em> in a per-<em>locus</em> way by looking at heatmaps.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_bw_heatmap.html">plot_bw_heatmap</a></span><span class="op">(</span></span>
<span>  <span class="va">h33_chip</span>,</span>
<span>  loci <span class="op">=</span> <span class="va">genes</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="plotting_files/figure-html/unnamed-chunk-16-1.png" width="672"></p>
<p>This gene annotation includes some variants of the same genes that
overlap a lot, so some of the rows look very similar or even like
duplicates.</p>
<p>Most of the parameters are shared between
<code>plot_bw_profile</code> and <code>plot_bw_heatmap</code> making it
possible to arrange them on top of each other later on, so it is
possible to align them in different ways:</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_bw_heatmap.html">plot_bw_heatmap</a></span><span class="op">(</span></span>
<span>  <span class="va">h33_chip</span>,</span>
<span>  loci <span class="op">=</span> <span class="va">genes</span>,</span>
<span>  mode <span class="op">=</span> <span class="st">"start"</span>,</span>
<span>  upstream <span class="op">=</span> <span class="fl">2500</span>,</span>
<span>  downstream <span class="op">=</span> <span class="fl">2500</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="plotting_files/figure-html/unnamed-chunk-17-1.png" width="672"></p>
<p>It is possible to specify <code>bin_size</code> to increase or
decrease resolution (this can also be done on profile plots):</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_bw_heatmap.html">plot_bw_heatmap</a></span><span class="op">(</span></span>
<span>  <span class="va">h33_chip</span>,</span>
<span>  loci <span class="op">=</span> <span class="va">genes</span>,</span>
<span>  mode <span class="op">=</span> <span class="st">"start"</span>,</span>
<span>  upstream <span class="op">=</span> <span class="fl">2500</span>,</span>
<span>  downstream <span class="op">=</span> <span class="fl">2500</span>,</span>
<span>  bin_size <span class="op">=</span> <span class="fl">20</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="plotting_files/figure-html/unnamed-chunk-18-1.png" width="672"></p>
<p>At this point, heatmap values are shown in order from maximum mean
value across all the beans to lowest.</p>
<p>By default, colorscale is set to percentiles 0.01 and 0.99 of whole
matrix range. You can set these values to fixed numbers. This is useful
mostly to plot together several heatmaps with the same color scale. For
example:</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_bw_heatmap.html">plot_bw_heatmap</a></span><span class="op">(</span></span>
<span>  <span class="va">h33_chip</span>,</span>
<span>  loci <span class="op">=</span> <span class="va">genes</span>,</span>
<span>  mode <span class="op">=</span> <span class="st">"start"</span>,</span>
<span>  upstream <span class="op">=</span> <span class="fl">2500</span>,</span>
<span>  downstream <span class="op">=</span> <span class="fl">2500</span>,</span>
<span>  bin_size <span class="op">=</span> <span class="fl">100</span>,</span>
<span>  zmin <span class="op">=</span> <span class="fl">2</span>,</span>
<span>  zmax <span class="op">=</span> <span class="fl">10</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="plotting_files/figure-html/unnamed-chunk-19-1.png" width="672"></p>
<p>It is also possible to change the color of the heatmap using any
<code>RColorBrewer</code> available palette:</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_bw_heatmap.html">plot_bw_heatmap</a></span><span class="op">(</span></span>
<span>  <span class="va">h33_chip</span>,</span>
<span>  loci <span class="op">=</span> <span class="va">genes</span>,</span>
<span>  bin_size <span class="op">=</span> <span class="fl">100</span>,</span>
<span>  cmap <span class="op">=</span> <span class="st">"RdYlBu"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="plotting_files/figure-html/unnamed-chunk-20-1.png" width="672"></p>
<p>There is also a <code>max_rows_allowed</code> parameter that defaults
to 10000. This allows to plot large matrices by binning vertically the
heatmap and plotting the underlying mean associated, provided that a
screen or a reasonable screen will not be able to display all the data
points anyway. If the number of rows is larger than
<code>max_rows_allowed</code>, rows will be binned in a way that the
resulting plotted matrix will have <code>max_rows_allowed</code> height
and a warning will be shown. Since the matrix is not subsampled only,
but the underlying values averaged, this may result in a smoothened
effect in the image. This is a similar approach to how image resizing
works.</p>
</div>
</div>
<div class="section level2">
<h2 id="plotting-signal-over-precalculated-granges">Plotting signal over precalculated GRanges<a class="anchor" aria-label="anchor" href="#plotting-signal-over-precalculated-granges"></a>
</h2>
<p>It is possible to plot the values already calculated in a
<code>GRanges</code> object by specifying the metadata columns used.
Essentially, anything calculated by <code>bw_loci</code> and
<code>bw_bins</code> functions can be passed to <code>plot_gr</code>
functions:</p>
<ul>
<li>
<code>plot_gr_violin</code>. Makes a violin plot of one or more
columns of a <code>GRanges</code> object.</li>
<li>
<code>plot_gr_scatter</code>. Makes a scatter plot of two different
columns.</li>
<li>
<code>plot_gr_density</code>. Makes a 2d histogram plot of two
different columns.</li>
</ul>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gr</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/bw_bins.html">bw_bins</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">h33_chip</span>, <span class="va">h3k9me3_chip</span><span class="op">)</span>,</span>
<span>  bin_size <span class="op">=</span> <span class="fl">2000</span>,</span>
<span>  genome <span class="op">=</span> <span class="st">"mm9"</span>,</span>
<span>  labels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"H33"</span>, <span class="st">"H3K9me3"</span><span class="op">)</span>,</span>
<span>  selection <span class="op">=</span> <span class="va">locus</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="../reference/plot_gr_scatter.html">plot_gr_scatter</a></span><span class="op">(</span><span class="va">gr</span>, </span>
<span>  x <span class="op">=</span> <span class="st">"H33"</span>,</span>
<span>  y <span class="op">=</span> <span class="st">"H3K9me3"</span>,</span>
<span>  remove_top <span class="op">=</span> <span class="fl">0.001</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="plotting_files/figure-html/unnamed-chunk-21-1.png" width="672"></p>
<p>This is useful if you are doing many pairwise combinations of
figures, or you want to do custom calculations on the bigWig values,
especially if the bin size is small and you are working with a large
amount of data points.</p>
</div>
<div class="section level2">
<h2 id="customizing-plots">Customizing plots<a class="anchor" aria-label="anchor" href="#customizing-plots"></a>
</h2>
<p>All plotting functions return a <code>ggplot</code> object. This
means that essentially you can overlay any other data you want or
further customize the results. For example:</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">profile_plot</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plot_bw_profile.html">plot_bw_profile</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">h33_chip</span>, <span class="va">h3k9me3_chip</span>, <span class="va">input_chip</span><span class="op">)</span>,</span>
<span>  loci <span class="op">=</span> <span class="va">genes</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">profile_plot</span> <span class="op">+</span> </span>
<span>  <span class="fu">ggtitle</span><span class="op">(</span><span class="st">"New title for the plot"</span><span class="op">)</span> <span class="op">+</span> <span class="co"># Add a custom title</span></span>
<span>  <span class="fu">geom_hline</span><span class="op">(</span>yintercept <span class="op">=</span> <span class="fl">1</span>, linetype<span class="op">=</span><span class="st">"dashed"</span><span class="op">)</span> <span class="co"># Add a horizontal line on y=1</span></span></code></pre></div>
<p><img src="plotting_files/figure-html/unnamed-chunk-22-1.png" width="672"></p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; R version 4.4.1 (2024-06-14)</span></span>
<span><span class="co">#&gt; Platform: x86_64-pc-linux-gnu</span></span>
<span><span class="co">#&gt; Running under: Ubuntu 22.04.5 LTS</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">#&gt; LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so;  LAPACK version 3.10.0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt;  [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8       </span></span>
<span><span class="co">#&gt;  [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8   </span></span>
<span><span class="co">#&gt;  [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C          </span></span>
<span><span class="co">#&gt; [10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C   </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; time zone: UTC</span></span>
<span><span class="co">#&gt; tzcode source: system (glibc)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] stats4    stats     graphics  grDevices utils     datasets  methods  </span></span>
<span><span class="co">#&gt; [8] base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt; [1] purrr_1.0.2          wigglescout_0.19.0   rtracklayer_1.64.0  </span></span>
<span><span class="co">#&gt; [4] GenomicRanges_1.56.1 GenomeInfoDb_1.40.1  IRanges_2.38.1      </span></span>
<span><span class="co">#&gt; [7] S4Vectors_0.42.1     BiocGenerics_0.50.0  ggplot2_3.5.1       </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;  [1] tidyselect_1.2.1            dplyr_1.1.4                </span></span>
<span><span class="co">#&gt;  [3] vipor_0.4.7                 farver_2.1.2               </span></span>
<span><span class="co">#&gt;  [5] Biostrings_2.72.1           bitops_1.0-9               </span></span>
<span><span class="co">#&gt;  [7] fastmap_1.2.0               RCurl_1.98-1.16            </span></span>
<span><span class="co">#&gt;  [9] GenomicAlignments_1.40.0    XML_3.99-0.17              </span></span>
<span><span class="co">#&gt; [11] digest_0.6.37               lifecycle_1.0.4            </span></span>
<span><span class="co">#&gt; [13] Cairo_1.6-2                 magrittr_2.0.3             </span></span>
<span><span class="co">#&gt; [15] compiler_4.4.1              rlang_1.1.4                </span></span>
<span><span class="co">#&gt; [17] sass_0.4.9                  tools_4.4.1                </span></span>
<span><span class="co">#&gt; [19] utf8_1.2.4                  yaml_2.3.10                </span></span>
<span><span class="co">#&gt; [21] knitr_1.48                  S4Arrays_1.4.1             </span></span>
<span><span class="co">#&gt; [23] labeling_0.4.3              curl_5.2.3                 </span></span>
<span><span class="co">#&gt; [25] DelayedArray_0.30.1         RColorBrewer_1.1-3         </span></span>
<span><span class="co">#&gt; [27] abind_1.4-8                 BiocParallel_1.38.0        </span></span>
<span><span class="co">#&gt; [29] withr_3.0.1                 desc_1.4.3                 </span></span>
<span><span class="co">#&gt; [31] grid_4.4.1                  fansi_1.0.6                </span></span>
<span><span class="co">#&gt; [33] colorspace_2.1-1            future_1.34.0              </span></span>
<span><span class="co">#&gt; [35] globals_0.16.3              scales_1.3.0               </span></span>
<span><span class="co">#&gt; [37] SummarizedExperiment_1.34.0 cli_3.6.3                  </span></span>
<span><span class="co">#&gt; [39] rmarkdown_2.28              crayon_1.5.3               </span></span>
<span><span class="co">#&gt; [41] ragg_1.3.3                  generics_0.1.3             </span></span>
<span><span class="co">#&gt; [43] httr_1.4.7                  rjson_0.2.23               </span></span>
<span><span class="co">#&gt; [45] ggbeeswarm_0.7.2            cachem_1.1.0               </span></span>
<span><span class="co">#&gt; [47] stringr_1.5.1               zlibbioc_1.50.0            </span></span>
<span><span class="co">#&gt; [49] parallel_4.4.1              ggrastr_1.0.2              </span></span>
<span><span class="co">#&gt; [51] XVector_0.44.0              restfulr_0.0.15            </span></span>
<span><span class="co">#&gt; [53] matrixStats_1.4.1           vctrs_0.6.5                </span></span>
<span><span class="co">#&gt; [55] Matrix_1.7-0                jsonlite_1.8.9             </span></span>
<span><span class="co">#&gt; [57] beeswarm_0.4.0              listenv_0.9.1              </span></span>
<span><span class="co">#&gt; [59] systemfonts_1.1.0           jquerylib_0.1.4            </span></span>
<span><span class="co">#&gt; [61] tidyr_1.3.1                 glue_1.8.0                 </span></span>
<span><span class="co">#&gt; [63] parallelly_1.38.0           pkgdown_2.1.1              </span></span>
<span><span class="co">#&gt; [65] codetools_0.2-20            stringi_1.8.4              </span></span>
<span><span class="co">#&gt; [67] gtable_0.3.5                BiocIO_1.14.0              </span></span>
<span><span class="co">#&gt; [69] UCSC.utils_1.0.0            munsell_0.5.1              </span></span>
<span><span class="co">#&gt; [71] tibble_3.2.1                pillar_1.9.0               </span></span>
<span><span class="co">#&gt; [73] furrr_0.3.1                 htmltools_0.5.8.1          </span></span>
<span><span class="co">#&gt; [75] GenomeInfoDbData_1.2.12     R6_2.5.1                   </span></span>
<span><span class="co">#&gt; [77] textshaping_0.4.0           evaluate_1.0.0             </span></span>
<span><span class="co">#&gt; [79] Biobase_2.64.0              lattice_0.22-6             </span></span>
<span><span class="co">#&gt; [81] highr_0.11                  Rsamtools_2.20.0           </span></span>
<span><span class="co">#&gt; [83] bslib_0.8.0                 SparseArray_1.4.8          </span></span>
<span><span class="co">#&gt; [85] xfun_0.48                   fs_1.6.4                   </span></span>
<span><span class="co">#&gt; [87] MatrixGenerics_1.16.0       pkgconfig_2.0.3</span></span></code></pre></div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Carmen Navarro.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>





  </body>
</html>
